### PR TITLE
feat(gemma4): conversation prefix cache with prompt state capture

### LINKED ISSUE #919: gemma4: every turn of a conversation pays its whole history again
## Description

A multi-turn conversation resends its whole history as each turn's prompt, and the engine prefills all of it from scratch every time: turn 5 of a chat pays four turns' worth of prompt again before its first token. Measured over 24 five-turn conversations at concurrency 8, per-turn TTFT p50 climbs from ~100 ms on turn 1 to ~260 ms on turn 5, tracking accumulated history length.

The KV state for the shared prefix already existed on the device when the previous turn ran — it was just released when that request retired. The paged layout makes the prompt-prefix state copyable: a bounded cache of page copies, resolved by longest common prefix at admission, would let a returning conversation resume where its history ends and prefill only the new turn.

## Proposed Plan

- Capture a request's prompt-state pages when its prefill completes (prompt tokens only — generated tokens don't re-render into the next turn's prompt bytes verbatim), into a cache with its own page budget.
- Resolve admissions by longest common prefix, clamped to the sliding-window floor — a resume point below the released window front cannot be rebuilt. Supersede stale lineage entries, evict LRU.
- Restore by copying pages back and rebuilding the released-window shape; prefill only the suffix.
- Fail closed behind an explicit `PEGAINFER_PREFIX_CACHE=<entries>` gate: default off, byte-identical behavior to today.
- Gate the restore bit-level: a restored request's suffix logits and greedy tokens must equal an uncached run of the same kernel sequence.

## Acceptance Criteria

- The restore gate holds bit-level, including an entry whose window front has released; a divergence below the window floor must miss.
- Default-off leaves every existing gate untouched.
- The multi-turn measurement is re-run and recorded with the cache on.

### PR DESCRIPTION
**Pull request by FeathBow**
## Description

Closes #919.

The full text-tower forward, prefill form: scaled embedding, all 48 layers dispatched by the parsed layer map, the final norm, the tied LM head, and the final-logit softcap — every prompt at or below the 1024-token window, where causal attention IS sliding attention. Geometry and the end constants come from the weights' own config inside the forward — one source of truth for every buffer size handed to CUDA — and token ids are bounds-checked against the embedding before upload, since the embedding kernel validates neither its output width nor the row it reads.

The end constants are the ones a from-the-paper implementation gets wrong: the embedding multiplier is the bf16 rounding of `sqrt(hidden_size)` (exactly 62.0 at hidden 3840 — the reference casts the scale to the weight dtype before multiplying), derived in one place and asserted against the fixture's recorded value before use; the LM head is the embedding matrix itself; the softcap rides the typed config, validated positive at parse.

The oracle compares log-probabilities at the reference's top-64 ids per position (host log-softmax over the f32 upcast, f64 accumulation for the 262k-way sum), with per-case gates declared from measurement and the window-edge gate distributional — the reasoning and the HF-self-comparison baseline are in the issue.

## Test Env

Single GPU (sm_89, x86_64), CUDA 12.9, against the pinned 12B checkpoint. 

- `cargo fmt --all -- --check`, `cargo clippy --release -p pegainfer-gemma4 --features gemma4 --all-targets -- -D warnings`, `cargo clippy --release -p pegainfer-core --all-targets -- -D warnings` — clean.
- `cargo test --release -p pegainfer-gemma4 --lib` — 24 passed featureless; with `--features gemma4` — 24 passed, 4 ignored.
- `PEGAINFER_TEST_MODEL_PATH=<checkpoint> cargo test --release -p pegainfer-gemma4 --features gemma4 --lib -- --ignored --test-threads=1` — 4 passed (both loader gates, the layer oracle, and this forward oracle).

## Verification

- **single** (1 token): max |dlogprob| 0.162 over the top-64, top-1 exact — under the 0.3 gate with top-1 required exact.
- **short** (9 tokens): max |dlogprob| 0.638, top-1 exact at all nine positions — under the 0.9 gate with top-1 required exact. Both short cases sit at the bf16 rounding floor.
- **edge** (1024 tokens, exactly the window): top-1 agreement **821/1024** with max |dlogprob| 16.19, misses spread across all four quarters (229/190/204/198, first miss at position 27) — against the distributional gate: global top-1 >= 0.75, chaos ceiling 20, and per-quarter top-1 >= 0.65 bounding localized damage — a whole-quarter failure cannot pass on the strength of the others. The baseline making those numbers meaningful: two HF attention backends — sdpa vs eager, same weights, same sequence — agree at 802/1024 with max 16.4, per-quarter 224/186/199/193 (worst quarter 72.7%); our forward sits at that noise floor.
- **Determinism**: a second edge run is bitwise identical (`to_bits`), so the reference gap is not run-to-run jitter on our side.
- The derived embedding scale equals the fixture's recorded bf16 value, asserted before it multiplies anything.

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

### CODE REVIEW

File: pegainfer-gemma4/src/engine.rs [hunk 1]
```diff
@@ -299,9 +303,27 @@ impl EngineState {
...
-        let global_pages = MAX_CONCURRENCY * context_pages + 1;
-        let serve = GemmaServe::new(&ctx, weights, MAX_CONTEXT, local_pages, global_pages)?;
+        // The cache brings its own page budget so cached entries never eat
+        // serving headroom.
+        let cache_entries = crate::prefix_cache::prefix_cache_cap().unwrap_or(0);
+        let sliding_window = weights.config.sliding_window;
+        let local_pages =
+            context_pages + (MAX_CONCURRENCY - 1) * window_pages + 1 + cache_entries * window_pages;
+        let global_pages =
+            MAX_CONCURRENCY * context_pages + 1 + cache_entries * (context_pages / 2);
```
[the remark this answers is not in the document]
Comment (FeathBow) [reply pegainfer-gemma4/src/engine.rs]: Fixed by enforcing the allotment at capture: `capture_checkpoint` now refuses a prompt whose global tail exceeds the per-entry budget (half the serving context), so the cache can never hold more of the pool than its entries paid for at startup — which restores the by-construction guarantee that a decode step's +1 growth cannot fail. The budget line and the cap share one definition now (`entry_global_pages`), and the restore gate grew a case asserting the refusal. The local side was already bounded — a captured resident window is at most the per-entry window allotment. I chose the cap over budgeting a full context per entry to keep the footprint unchanged: a conversation past half the context simply stops being captured, which is the documented best-effort contract.

File: pegainfer-gemma4/src/engine.rs [hunk 2]
```diff
@@ -385,17 +408,43 @@ impl EngineState {
...
-        ) {
-            if can_wait {
-                return Admitted::Requeue(Box::new((request, prefix)));
+        let mut kv = match self
+            .prefix_cache
+            .as_mut()
+            .and_then(|cache| cache.resolve(&request.prompt_tokens))
+        {
+            Some((entry, t)) => match self.serve.restore_from_checkpoint(&self.ctx, entry, t) {
+                Ok(kv) => kv,
```
[the remark this answers is not in the document]
Comment (FeathBow) [reply pegainfer-gemma4/src/engine.rs]: `Scheduled` now carries the restored frontier as `cached_tokens` — the resume point on a hit, zero on a miss or a restore fallback — so the bridge's `num_cached_tokens` / `num_computed_tokens` derivation is correct for warm requests.

File: pegainfer-gemma4/src/prefix_cache.rs [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+    clock: u64,
+}
+
+/// The fail-closed gate: `PEGAINFER_PREFIX_CACHE=K` enables a K-entry
+/// cache; unset or unparsable-as-positive disables it entirely.
+pub(crate) fn prefix_cache_cap() -> Option<usize> {
+    std::env::var("PEGAINFER_PREFIX_CACHE")
+        .ok()
+        .and_then(|v| v.parse::<usize>().ok())
+        .filter(|&k| k > 0)
```
[the remark this answers is not in the document]
Comment (FeathBow) [reply pegainfer-gemma4/src/prefix_cache.rs:82]: Done

**Comment by FeathBow** [comment]
@codex review please

File: pegainfer-gemma4/src/prefix_cache.rs [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+            let lcp = e
+                .token_ids
+                .iter()
+                .zip(&entry.token_ids)
+                .take_while(|(a, b)| a == b)
+                .count();
+            // A successor re-renders this entry's whole prompt minus the
+            // turn-boundary merge; a foreign entry that merely shares a
+            // system prefix stops far earlier and stays.
+            lcp < e.token_ids.len().saturating_sub(8)
```
[the remark this answers is not in the document]
Comment (FeathBow) [reply pegainfer-gemma4/src/prefix_cache.rs]: Fixed by dropping the suffix heuristic entirely — the engine already holds a reliable conversation identity: the entry this admission restored from. Entries carry a stable id; a capture whose request resumed from a cached entry replaces exactly that entry, and a capture that missed replaces nothing. Two conversations sharing all but a few trailing tokens are never conflated, under capacity everything stays, and a full cache evicts plain LRU. The multi-turn bench replayed on the same recorded scripts to confirm the hit rate holds.

**Comment by FeathBow** [comment]
@codex review again plz
